src/wpool/mongoose_gun_worker.erl

```diff
@@ -0,0 +1,101 @@
+-module(mongoose_gun_worker).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/2]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2]).
+
+-record(state, {host :: inet:hostname() | inet:ip_address(),
+                port :: inet:port_number(),
+                opts :: gun:opts(),
+                pid :: pid() | undefined,
+                monitor :: reference() | undefined}).
+
+-type request() :: {request,
+                    Path::iodata(),
+                    Method::iodata(),
+                    Headers::gun:headers(),
+                    Query::iodata(),
+                    Retries::pos_integer(),
+                    Timeout::non_neg_integer()}.
+
+-include("mongoose_logger.hrl").
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+start_link(Destination, Options) ->
+    gen_server:start_link(?MODULE, {Destination, Options}, []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+-spec init({{inet:hostname() | inet:ip_address(), inet:port_number()}, gun:opts()}) ->
+    {ok, State :: #state{}}.
+init({{Host, Port}, Opts}) ->
+    {ok, #state{host = Host, port = Port, opts = maps:merge(default_opts(), Opts)}}.
+
+-spec handle_call(Request :: request(), From :: {pid(), Tag :: term()}, State :: #state{}) ->
+                     {reply, Reply :: term(), NewState :: #state{}}.
+handle_call({request, FullPath, Method, Headers, Query, Retries, Timeout}, _From, State) ->
+    LHeaders = lowercase_headers(Headers),
+    NewState = create_or_get_connection(State),
+    Res = send_request(FullPath, Method, LHeaders, Query, Retries + 1, Timeout, NewState#state.pid),
```

Comment:
HTTP/2 is a fully-duplex enabled protocol, yet here we block the connection process until we get response, locking other client from sending request. This should never happen - the response should be handled asynchronously (that's why gun send a message on completion!) and then the reply to calling client should be made.

Reply:
Btw. this should happen only if HTTP/2 protocol is negotiated - please remember that `gun` falls back to HTTP 1.1 unless HTTP/2 is forced with options. Unless `gun` already internally queues up requests when HTTP 1.1 is used - that I simply don't know.